### DISCUSSION TITLE
In FileElement, how does a file go from temporarily to permanently uploaded?

### DISCUSSION THREAD

**Discussion by costa-mat** [vueform/vueform#115] [Questions] [answered]
I have used the FileElement option [uploadTempEndpoint](https://vueform.com/reference/file-element#option-upload-temp-endpoint) to upload a file in a form. I want to make the file name [clickable](https://vueform.com/reference/file-element#option-clickable), but I can't do this while the file status is temporarily uploaded.

I would like to know 
1) when does the file stage go from 'file temporarily uploaded' to 'file permanently uploaded', 
2) and if I can tell a specific file element to consider its file as permanently uploaded, for exemple by accessing its el$ and calling a suitable method (FileElement has a property [stage](https://vueform.com/reference/file-element#property-stage), but it is a computed property.)

I would like for the user to be able to click and download the file that they have just uploaded, so they can check that the file is correct.

In this section [clickable](https://vueform.com/reference/file-element#option-clickable), there is the information:
'The file is only considered uploaded once its value is a string that contains the filename.' 
Do I have to directly access this file object and change this value to a string? Or is this done as a consequence of some method? How can I access this file object to change its value to a string? And how does the FileElement will know that this change took place?

I have attempted doing something like this
```
<!-- Using a function to upload file -->
<FileElement :endpoint="async function(file, el$){
  let data = el$.form$.convertFormData({
    file,
  }) // befomes `FormData` object

  /* upload file */
  // I ATTEMPTED THIS
  $el.stage = 3
  // OR THIS
  $el.value = 'name of file'
  return {
    tmp: '...', // the temp file identifier
    originalName: '...', // the original name of the file that will be displayed to the user
  }
}" ... />
```

And this caused the file name to became temporarily clickable, but since both stage and value are computed properties, this gets reverted back a few seconds later.

**Comment by adamberecz** [comment, marked as answer]
The stage is determined automatically based on the type of the value:
- `null` - stage `0` (empty)
- `File` type - stage `1` (file selected)
- `object` type - stage `2` (temp uploaded)
- `string` type - stage `3` (uploaded)

If you return a **string** from `:upload-temp-endpoint` the file will be considered uploaded and will be clickable.

```html
<!-- Using a function to upload file -->
<FileElement :upload-temp-endpoint="async function(file, el$){
  let data = el$.form$.convertFormData({
    file,
  }) // befomes `FormData` object

  // ...

  return 'filename.jpg'
}" ... />
```

The docs contained an error calling the prop `:endpoint` instead of `:upload-temp-endpoint`. It got fixed and I also added an [example](https://vueform.com/reference/file-element#option-upload-temp-endpoint) of skipping the temp upload.

**Reply by costa-mat** [reply]
Thanks a lot, this solved my problem
